File: amber/src/main/python/core/architecture/handlers/control/update_executor_handler.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from core.architecture.handlers.control.control_handler_base import ControlHandler
from proto.org.apache.texera.amber.engine.architecture.rpc import (
    EmptyReturn,
    UpdateExecutorRequest,
)
from core.util import get_one_of
from proto.org.apache.texera.amber.core import OpExecWithCode


class UpdateExecutorHandler(ControlHandler):
    async def update_executor(self, req: UpdateExecutorRequest) -> EmptyReturn:
        op_exec_with_code: OpExecWithCode = get_one_of(req.new_exec_init_info)
        self.context.executor_manager.update_executor(
            op_exec_with_code.code, self.context.executor_manager.executor.is_source
        )
        return EmptyReturn()

File: amber/src/main/scala/org/apache/texera/amber/engine/architecture/worker/promisehandlers/UpdateExecutorHandler.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.texera.amber.engine.architecture.worker.promisehandlers

import com.twitter.util.Future
import org.apache.texera.amber.engine.architecture.rpc.controlcommands.{
  AsyncRPCContext,
  UpdateExecutorRequest
}
import org.apache.texera.amber.engine.architecture.rpc.controlreturns.EmptyReturn
import org.apache.texera.amber.engine.architecture.worker.DataProcessorRPCHandlerInitializer
import org.apache.texera.amber.util.VirtualIdentityUtils

trait UpdateExecutorHandler {
  this: DataProcessorRPCHandlerInitializer =>

  override def updateExecutor(
      request: UpdateExecutorRequest,
      ctx: AsyncRPCContext
  ): Future[EmptyReturn] = {
    val workerIdx = VirtualIdentityUtils.getWorkerIndex(actorId)
    // Close the existing executor (if any) before replacing it to avoid resource leaks.
    val oldExecutor = dp.executor
    if (oldExecutor != null) {
      oldExecutor.close()
    }
    setupExecutor(request.newExecInitInfo, workerIdx, cachedTotalWorkerCount)
    dp.executor.open()
    EmptyReturn()
  }

}
